### ISSUE TITLE
Expections in definitions of methods handled by EventDispatcher.dispatch_event are being passed silently

### ISSUE THREAD

**Issue by The6apcuk** [pythonarcade/arcade#547]
## Bug Report
Expections in definitions of methods (e.g. on_key_press, on_key_release, on_draw) handled by EventDispatcher.dispatch_event are being passed silently. As far as I understand it is done to ignore the not defined attributes for user's defined class (e.g on_mouse_leave, on_expose, on_activate etc.) But in this case if a real mistake related to attribute fetching was done in the code, it is hard to understand what is going on.

### Actual behavior:
With next code the application runs without any expection, even though int variable "a" declared in definition of on_draw method doesn't have an attribute b. 

```
import arcade

class MyGame(arcade.Window):
    """
    Main application class.
    """

    def __init__(self):
        # Call the parent class and set up the window
        self.debug=True
        super().__init__(1000, 600, 'SCREEN_TITLE')

    def setup(self):
     pass
    def on_draw(self):
        """ Render the screen. """
        a = 5
        a.b
        # Clear the screen to the background color
        arcade.start_render()

    def on_key_press(self, key, modifiers):
        a = 5
        a.b

    def on_key_release(self, key, modifiers):
        a = 5
        a.b
    def on_update(self, delta_time):
        pass



def main():
    """ Main method """
    window = MyGame()
    window.setup()
    arcade.run()


if __name__ == "__main__":
    main()
```
In case if in dispatch_event method of class EventDispatcher of pyglet/event.py module is modified, the expetion can be seen:
```
        # Check instance for an event handler
        try:
            if getattr(self, event_type)(*args):
                return EVENT_HANDLED
        except AttributeError as exept:
            if self.debug:
                logging.error(f'game error {exept}')
```

### Expected behavior:
Attribute errors should not pass silently 

### Steps to reproduce/example code:
Execute code in Actual results

### What should be added/changed?
Change exception handling in the dispatch_event method of class EventDispatcher of pyglet/event.py module to raise more specific error in case of missing attribute. 
From (current code):
```
        # Check instance for an event handler
        try:
            if getattr(self, event_type)(*args):
                return EVENT_HANDLED
        except AttributeError:
            pass
```
to next code, in case if EventException is sutible for this case, otherewise a more specific exception should be created and inhereted from AttributeError:
```
        try:
            if not hasattr(self, event_type):
                raise EventException(f'Failed to fetch {event_type} from {self}')
            elif getattr(self, event_type)(*args):
                return EVENT_HANDLED
        except EventException:
            pass
```
### What would it help with?
In this case exceptions for real mistakes in the code are not passed silently

**Comment by pvcraven** [comment]
I think this might be a Pyglet-level issue. More research is needed.

**Comment by pvcraven** [comment]
See Pyglet issues:
* https://github.com/pyglet/pyglet/issues/105
* https://github.com/pyglet/pyglet/issues/87

**Comment by specht** [comment]
Correct me if I'm wrong, but I think there's never an excuse to swallow an exception, and especially so if you call user-defined callback methods. And for a library which is targeted at students, no less... how are they supposed to find what's wrong with their code if attribute errors are just being swallowed? So it's a problem in both arcade and pyglet and it has to be fixed in both libraries. Can't we just test if the handler is defined and call it if it is without an catch-pass blocks?

**Comment by pvcraven** [comment]
Arcade, Pyglet and similar libraries are written by people who donate thousands of their own hours so that people, such as yourself, can enjoy programming with them. There's no paid positions, no QA testers, no professional documentation people. Just people who have a choice to spend some extra time testing, or sit with their kids at a soccer game. And sometimes the kids come first. So yes, there is an excuse.

If you are unhappy a bug exists or feature is missing, don't get upset with other people, step up and get involved! With Python and open-source it is pretty straight-forward to find the code that is causing the issue. Or if you can't, start learning how by asking other people. Then you can create your own solution and generate a pull-request. Then thousands of people can appreciate your work.

Although as a first step, it is a good idea to check to make sure someone else isn't working on it already. In this case, the linked Pyglet bug report already has a pull-request with  code to solve the problem, and it should be out in the next release of Pyglet. If you want, you can pull that code and contribute by testing it! You can also always go to that bug and ask about timelines for a release, and it isn't a bad idea to drop a nice word or two to the maintainers.

**Comment by specht** [comment]
Hey, I'm sorry I upset you. Reading your reply, I clearly have come across in a way which I didn't intend. First of all, thanks for all the work you're doing! I know all of this, I have kids too, and I'm also involved in open source, and I have zero expectations towards anyone and I'm grateful for all the good work that's being done. It's a given for me. Sorry I came across as demanding and entitled.

I was trying (and failing) to make a purely technical argument, because like this example shows, swallowing exceptions is a dangerous thing. But you're completely right, I'm talking too much, let me shoot you some pull requests instead. I was trying to make a point that it's not just a Pyglet problem, but the same problem exists in Arcade in 26 places, and I wanted to hear your opinion on the fix that's been proposed above, because you're the project lead.

So: 1) sorry and 2) thanks, and 3) I'll get busy now. :-D

**Comment by pvcraven** [comment]
Update to Pyglet 1.5.0 and it should be fixed. Please re-open if not. Working for me anyway.

**Comment by pvcraven** [comment]
Some events are still being passed silently, depending on the method.

**Comment by einarf** [comment]
I think `AttributeError` and/or `KeyError` gets consumed in `on_draw` at least. I had this issue in the `shader_experimental` branch.

**Comment by dogwynn** [comment]
Another "fix":

```python3
import functools
import logging

def traceback(method):
    @functools.wraps(method)
    def catcher(*a, **kw):
        try:
            return method(*a, **kw)
        except Exception:
            logging.exception('Exception caught:')
    return catcher
```

Use the above as a decorator for `on_update` and `on_draw` methods of Window class.

**Comment by einarf** [comment]
This was fixed in pyglet 1.5.4. Also arcade views should not longer consume attributeerror,
